### DISCUSSION TITLE
What's the correct way to programatically create files?

### DISCUSSION THREAD

**Discussion by TheCodeLamp** [twpayne/chezmoi#3333] [Q&A]
I have a number of distroboxes on my machine and I would like to create a zellij layout for each of them.
I can get the names in a list using the "dostrobox-list" command, and some parsing. The problems arrive when I want to create the files. I know that I can use a script to just generate the files, but I feel like it would be better if I could use the templateing system, or at least something so that chezmoi can keep track of file changes.

**Comment by twpayne** [comment]
If I understand correctly, distrobox shares your home directory across all your boxes, and [zellij uses a single config file](https://zellij.dev/documentation/configuration.html), so you only need to create `dot_config/zellij/config.kdl` in your chezmoi source directory. chezmoi will then write this to `~/.config/zellij/config.kdl` in your home directory and all your distroboxes will find it there. Or is your setup different to the standard setup?

**Reply by TheCodeLamp** [reply]
What I want is to be able to write `zellij --layout feodra` where it then uses a layout I have saved called `~/.config/zellij/layouts/fedora.kdl`. That layout automatically does some commands and puts me inside the distrobox. But to do this I need a layout file in the layouts folder. I have a script and a template in the .chezmoitemplates folder that I can just supply a variable to to create the right file. `{{- template "distrobox.kdl" "fedora" -}}` to create the contents of the fedora file.

However I don't know how to create the actual file using chezmoi. Right now I use a script that just executes `chezmoi execute-template $"{{- template \"distrobox-zellij.kdl\" \"($box)\" -}}"` where the `box` var is the name of the box. It works but it doesn't keep track of the created file.

If this is the best way to do this it's fine, I was just wondering if there was a way to create files using the template system.

**Reply by twpayne** [reply]
Thank you for the details.

So, with my updated understanding:

1) All distroboxes share the same home directory, so making `~/.config/zellij/config.kdl` a chezmoi template with a `layout` section dependent on the distro name is not an option.
2) zellij's config files do not provide an option to switch layouts depending on the distro, so you need a separate file for each layout.
3) You specify the distro manually with the `--layout $DISTRO` flag to zellij.
4) There are multiple distros, so you want to create a `~/.config/zellij/layouts/$DISTRO.kdl` file for each `$DISTRO` in the output of `distrobox-list`.

If the above four statements are true what do you think about the following?

* Instead of running `zellij --layout fedora`, can you run `zellij --layout $MY_PREFERRED_LAYOUT` instead, where `$MY_PREFERRED_LAYOUT` depends on the distribution instead? For example, Fedora, RedHat, CentOS, RHEL, etc. presumably are quite similar, as are Debian, Ubuntu, Pop!_OS, etc. You can use chezmoi templates to create an alias in your shell config, e.g. `alias zellij=zellij --layout {{ .chezmoi.osRelease.idLike }}`.
* You now only need one file in `~/.config/zellij/layouts` for each style of Linux, so you no longer need one per line of output of `distobox-list`. Presumably the number of styles of Linux is relatively small, so you no longer need to generate files.

**Reply by twpayne** [reply]
Did this answer your question?
